# ManyValuedContext

This notebook demonstrates the basic functionality of ``MVContext`` class.

The class allows one to create a many valued context from FCA theory. Set its objects and descriptions of these objects. Then one can compute intention of subset of objects and extension of subset of attributes.

The data can be loaded and save to json

## Create MVContext

In this example, we create a many valued context from California Housing dataset.

In [1]:
#load data from sci-kit learn
from sklearn.datasets import fetch_california_housing
california_data = fetch_california_housing(as_frame=True)
df = california_data['data'].round(3)

We define a specific type of PatternStructure for each column of a dataframe and create a MVContext

In [16]:
from fcapy.mvcontext import MVContext, PS
ptypes = {f: PS.IntervalPS for f in df.columns}
K = MVContext(
    df.values, pattern_types=ptypes,
    object_names=('house_'+df.index.astype(str)).tolist(),
    attribute_names=df.columns.tolist(),
)

Get the description of the first house

In [17]:
K.intention_i([0])

{0: (8.325, 8.325),
 1: (41.0, 41.0),
 2: (6.984, 6.984),
 3: (1.024, 1.024),
 4: (322.0, 322.0),
 5: (2.556, 2.556),
 6: (37.88, 37.88),
 7: (-122.23, -122.23)}

The output says the following:
* value of the feature $0$ lies in interval $[8.325, 8.325]$,
* value of the feature $1$ lies in interval $[41.0, 41.0]$,
* etc

Get the common description of houses {``house_1``, ``house_42``}

In [20]:
K.intention(['house_1', 'house_42'])

{'MedInc': (1.025, 8.301),
 'HouseAge': (21.0, 49.0),
 'AveRooms': (3.772, 6.238),
 'AveBedrms': (0.972, 1.069),
 'Population': (462.0, 2401.0),
 'AveOccup': (2.11, 2.444),
 'Latitude': (37.84, 37.86),
 'Longitude': (-122.26, -122.22)}

The output says the following:
* the common value of the feature 'MedInc' in interval $[1.025, 8.301]$,
* the common value of the feature 'HouseAge' lies in interval $[21.0, 49.0]$,
* etc

Get a number of houses with an age in a closed interval [10, 21]

In [4]:
len(K.extension({'HouseAge': (10, 21)}))

5434

## Read/write MVContext to json

Once the work is done one can write the context to a json file

In [5]:
path = "example.json"
K.write_json(path)
K_new = MVContext.read_json(path)
assert K_new == K

!rm {path}